Strawman for vendor plugins.

markstory opened the pull request:
This is a basic strawman for changing how plugins are installed. Instead of putting composer managed plugins into /plugins, this will allow a path configuration file to define where plugins are located. Having this configuration file, will allow the plugin-installer to update the file.

This will require additional changes in cakephp/app, and cakephp/plugin-installer.

Strawman for simpler approach to cakephp/app#206

markstory committed:
Strawman for vendor plugins.

This is a basic strawman for changing how plugins are installed. Instead
of putting composer managed plugins into /plugins, this will allow
a path configuration file to define where plugins are located. Having
this configuration file, will allow the plugin-installer to update the
file.

This will require additional changes in cakephp/app, and
cakephp/plugin-installer.

josegonzalez commented:
Why not allow that path to be configured in the composer.json?

ADmad commented:
@josegonzalez How would that help?

josegonzalez commented:
Less cakephp-specific config.

ADmad commented on src/Core/Plugin.php:
```diff
@@ -170,6 +184,24 @@ public static function load($plugin, array $config = [])
+            static::$_config = [];
+            return;
+        }
+        $config = include CONFIG . 'plugins.php';
```
Instead of using `include` how about using `Configure::load()` and also adding a methods to update and dump the map?  That way all the code for maintaining the map remains in core and the plugin-installer would just need to call the appropriate method upon plugin installation / removal.

AD7six replied:
`Configure::load` does seem a little better. 

The only thing I don't like about this is we're storing a config file for "this name is in this location" and `composer.json` also does "this name(space) is in this location" - but to a certain extent it becomes a choice of which is the lesser evil.

ADmad replied:
@AD7six Yes it is some duplication. But using a new config file is easier that using composer to get that info.

markstory replied:
Good points I will update the code to use the PhpConfig class to read the file.

markstory replied:
@AD7six @josegonzalez Does composer have a way to query which packages arr installed and what their types are?

ADmad replied:
`composer show -i` gives you listed of installed packages and `composer show <package-name>` gives you info about the package which includes the package type. But these commands seem to take quite some time to show output. So even if we were to use them we would have to cache the info, so I would rather just directly use a config file.

AD7six replied:
I don't think so.

There's `vendor/composer/installed.json` which does list all installed packages and their type - but that's an implementation detail of composer itself.

ceeram replied:
Would also mean you need composer/composer itself as require in
cakephp/cakephp bringing in a whole bunch of other packages as well. Also
would mean you dont only need composer to install but also to run cakephp.
Op 12 jan. 2015 13:53 schreef "Andy Dawson" <notifications@github.com>:

> In src/Core/Plugin.php
> <https://github.com/cakephp/cakephp/pull/5641#discussion-diff-22787899>:
>
> > @@ -170,6 +184,24 @@ public static function load($plugin, array $config = [])
> >      }
> >
> >      /**
> > +     * Loads the config mappings for plugins.
> > +     *
> > +     * The map data in CONFIG/plugins.php map data is used set default paths
> > +     * for where plugins are located.
> > +     *
> > +     * @return void
> > +     */
> > +    protected static function _loadConfigMap()
> > +    {
> > +        if (!file_exists(CONFIG . 'plugins.php')) {
> > +            static::$_config = [];
> > +            return;
> > +        }
> > +        $config = include CONFIG . 'plugins.php';
>
> I don't think so.
>
> There's vendor/composer/installed.json which does list all installed
> packages and their type - but that's an implementation detail of composer
> itself.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cakephp/cakephp/pull/5641/files#r22787899>.
>

markstory replied:
Yeah, I'd rather not invoke all of composer on each request.

markstory commented:
Is this looking reasonable enough for me to go ahead and make the required changes in cakephp/app and cakephp/plugin-installers?

AD7six commented:
@markstory think so :+1:

ADmad commented:
@markstory You don't mean to update the code to use Configure::load() and add methods to update and dump config as i suggested?

markstory commented:
I guess using `Configure` would the mean Plugin wouldn't need to store that configuration data itself which is good in a way.

markstory committed:
Use Configure to load plugin paths.

Switch to using more standard framework features to load the plugin
configuration file. This will also allow us to dump() to export files
from the plugin-installer.

markstory committed:
Remove sample config file from cakephp tests.

Not having the config file ensures apps work without a config file.

markstory committed:
Add tests for Plugin path config.

ADmad commented on src/Core/Plugin.php:
```diff
@@ -119,6 +120,13 @@ public static function load($plugin, array $config = [])
             return;
         }
 
+        if (!Configure::check('pluginPaths')) {
```
Here you check for key `pluginPaths` but below `'plugins.' . $plugin` is used for reading.

markstory replied:
derp.

ADmad commented on tests/TestCase/Core/PluginTest.php:
```diff
@@ -120,6 +120,19 @@ public function testLoadSingleWithBootstrapAndRoutes()
+        $this->assertEquals(APP . 'src' . DS, Plugin::classPath('TestPlugin'));
+    }
+
+
```
Extra new line.

markstory committed:
Fix a few mistakes.

AD7six commented:
This looks good to merge, any other input?

markstory committed:
Plugin::loadAll() should also read the config file.

When using loadAll() to load plugins, plugins defined in the path config
file should be loaded.

markstory commented:
I also updated `Plugin::loadAll()` to use the config file.

markstory committed:
Add missing doc block.
